<launch>
    <!-- Arguments -->
    <arg name="use_sim_time" default="true" />
    <arg name="map" default="$(find-pkg-share my_robot_bringup)/maps/test_world.yaml" />
    <arg name="params_file" default="$(find-pkg-share my_robot_bringup)/config/nav2_params.yaml" />
    
    <!-- Robot URDF path -->
    <let name="urdf_path" 
        value="$(find-pkg-share my_robot_description)/urdf/my_robot.urdf.xacro" />
    
    <!-- Config paths -->
    <let name="gazebo_bridge_config" 
        value="$(find-pkg-share my_robot_bringup)/config/gazebo_bridge.yaml" />
    <let name="rviz_config" 
        value="$(find-pkg-share my_robot_bringup)/config/nav2_rviz.rviz" />

    <!-- Robot State Publisher -->
    <node pkg="robot_state_publisher" exec="robot_state_publisher">
        <param name="robot_description" value="$(command 'xacro $(var urdf_path)')" />
        <param name="use_sim_time" value="$(var use_sim_time)" />
    </node>

    <!-- Launch Gazebo with world -->
    <include file="$(find-pkg-share ros_gz_sim)/launch/gz_sim.launch.py">
        <arg name="gz_args" value="$(find-pkg-share my_robot_bringup)/worlds/test_world.sdf -r" />
    </include>

    <!-- Spawn robot in Gazebo at origin -->
    <node pkg="ros_gz_sim" exec="create" args="-topic robot_description -x 0 -y 0 -z 0.1" />

    <!-- Gazebo-ROS bridge -->
    <node pkg="ros_gz_bridge" exec="parameter_bridge">
        <param name="config_file" value="$(var gazebo_bridge_config)" />
    </node>

    <!-- Scan Timestamp Fixer - republishes /scan_raw to /scan with corrected timestamps -->
    <node pkg="my_robot_bringup" exec="scan_timestamp_fixer.py" name="scan_timestamp_fixer">
        <param name="use_sim_time" value="$(var use_sim_time)" />
    </node>

    <!-- Odom to TF broadcaster (publishes odom -> base_footprint TF from /odom topic) -->
    <node pkg="my_robot_bringup" exec="odom_to_tf.py" name="odom_to_tf">
        <param name="use_sim_time" value="$(var use_sim_time)" />
    </node>

    <!-- TF Republisher - fixes static transform timestamp issues in slow simulations -->
    <node pkg="my_robot_bringup" exec="tf_republisher.py" name="tf_republisher">
        <param name="use_sim_time" value="$(var use_sim_time)" />
    </node>

    <!-- Nav2 bringup - both navigation and localization together -->
    <include file="$(find-pkg-share nav2_bringup)/launch/bringup_launch.py">
        <arg name="use_sim_time" value="$(var use_sim_time)" />
        <arg name="map" value="$(var map)" />
        <arg name="params_file" value="$(var params_file)" />
        <arg name="autostart" value="true" />
    </include>

    <!-- RViz with our custom config -->
    <node pkg="rviz2" exec="rviz2" output="screen" args="-d $(var rviz_config)">
        <param name="use_sim_time" value="$(var use_sim_time)" />
    </node>

    <!-- Navigation FSM - manages states and recovery behaviors -->
    <node pkg="my_robot_bringup" exec="navigation_fsm.py" name="navigation_fsm">
        <param name="use_sim_time" value="$(var use_sim_time)" />
        <param name="recovery_timeout" value="10.0" />
        <param name="max_replan_attempts" value="3" />
        <param name="stuck_threshold" value="5.0" />
    </node>

</launch>
